curl/src/main/scala/org/http4s/curl/websocket/Connection.scala

```diff
@@ -0,0 +1,230 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.curl.websocket
+
+import cats.effect.IO
+import cats.effect.Resource
+import cats.effect.SyncIO
+import cats.effect.implicits._
+import cats.effect.kernel.Deferred
+import cats.effect.kernel.Ref
+import cats.effect.std.Dispatcher
+import cats.effect.std.Queue
+import cats.implicits._
+import org.http4s.client.websocket._
+import org.http4s.curl.internal.Utils
+import org.http4s.curl.internal.Utils.throwOnError
+import org.http4s.curl.unsafe.libcurl
+import org.http4s.curl.unsafe.libcurl_const
+import scodec.bits.ByteVector
+
+import scala.scalanative.unsafe._
+import scala.scalanative.unsigned._
+
+import Connection._
+
+final private class Connection private (
+    val handler: Ptr[libcurl.CURL],
+    receivedQ: Queue[IO, Option[WSFrame]],
+    receiving: Ref[SyncIO, Option[Receiving]],
+    established: Deferred[IO, Unit],
+    dispatcher: Dispatcher[IO],
+    breaker: Breaker,
+) {
+
+  /** received frames */
+  def receive: IO[Option[WSFrame]] =
+    receivedQ.take <* breaker.feed
+
+  private def enqueue(wsframe: WSFrame): Unit =
+    dispatcher.unsafeRunAndForget(
+      breaker.drain >> receivedQ.offer(wsframe.some)
+    )
+
+  /** libcurl write callback */
+  def onReceive(
+      buffer: Ptr[CChar],
+      size: CSize,
+      nmemb: CSize,
+  ): CSize = {
+    val realsize = size * nmemb
+    val meta = libcurl.curl_easy_ws_meta(handler)
+
+    val toEnq = receiving
+      .modify {
+        case Some(value) =>
+          val next = value.add(buffer, realsize)
+
+          val optF = next.toFrame
+          if (optF.isDefined) (None, optF)
+          else (Some(next), None)
+        case None =>
+          if (meta.isClose) {
+            (None, Some(WSFrame.Close(200, "Closed by server")))
+          } else {
+            val frameType =
+              if (meta.isText) ReceivingType.Text
+              else if (meta.isBinary) ReceivingType.Binary
+              else if (meta.isPing) ReceivingType.Ping
+              else throw InvalidFrame
+
+            val recv =
+              Receiving(buffer, realsize, meta.isFinal, frameType, meta.bytesLeft.toULong)
+
+            val optF = recv.toFrame
+            if (optF.isDefined) (None, optF)
+            else (Some(recv), None)
+          }
+      }
+      .unsafeRunSync()
+
+    toEnq.foreach(enqueue)
+
+    realsize
+  }
+
+  def onEstablished(): Unit = dispatcher.unsafeRunAndForget(established.complete(()))
+
+  def onTerminated(result: Either[Throwable, Unit]): Unit = {
+    dispatcher.unsafeRunAndForget(receivedQ.offer(None))
+    result match {
+      case Left(err) =>
+        Console.err.println("Websocket connection terminated!")
+        err.printStackTrace(Console.err)
+      case _ => ()
+    }
+  }
```

Comment:
I think if you do
```scala
dispatcher.unsafeRunAndForget(
  receivedQ.offer(None) *> IO.fromEither(result)
)
```

Then the error will be reported through the Cats Effect unhandled error mechanisms.